cli/popper/runner_slurm.py

```diff
@@ -0,0 +1,206 @@
+import os
+import tempfile
+import subprocess
+
+from popper import utils as pu
+from popper import scm
+from popper.cli import log as log
+from popper.runner import StepRunner as StepRunner
+from popper.runner_host import DockerRunner as HostDockerRunner
+from popper.runner_host import HostRunner
+
+
+class SlurmRunner(StepRunner):
+
+    spawned_jobs = set()
+
+    def __init__(self, config):
+        super(SlurmRunner, self).__init__(config)
+
+    def __exit__(self, exc_type, exc, traceback):
+        SlurmRunner.spawned_jobs = set()
+
+    def generate_script(self, cmd, job_id, job_script):
+        with open(job_script, "w") as f:
+            f.write("#!/bin/bash\n")
+            f.write(cmd)
+
+    def submit_batch_job(self, cmd, step):
+        job_id = pu.sanitized_name(step['name'], self.config.wid)
+        temp_dir = "/tmp/popper/slurm/"
+        os.makedirs(temp_dir, exist_ok=True)
+        job_script = os.path.join(temp_dir, f"{job_id}.sh")
+        out_file = os.path.join(temp_dir, f"{job_id}.out")
+        err_file = os.path.join(temp_dir, f"{job_id}.err")
+
+        self.generate_script(cmd, job_id, job_script)
+
+        sbatch_cmd = "sbatch --wait "
+        sbatch_cmd += f"--job-name {job_id} "
+        sbatch_cmd += f"--output {out_file} "
+        sbatch_cmd += f"--error {err_file} "
+
+        if hasattr(self.config, 'resman_options'):
+            options = self.config.resman_options
+            if options.get(step['name']):
+                step_slurm_config = options[step['name']]
+                for config_key, config_val in step_slurm_config.items():
+                    if isinstance(config_val, bool):
+                        if len(config_key) == 1:
+                            sbatch_cmd += f"-{config_key} "
+                        else:
+                            sbatch_cmd += f"--{config_key} "
+                    else:
+                        if len(config_key) == 1:
+                            sbatch_cmd += f"-{config_key} {config_val} "
+                        else:
+                            sbatch_cmd += f"--{config_key} {config_val} "
```

Comment:
since boolean values can be assigned true/false, another way to write this that saves a couple of lines and indentation levels:

```python
resman_options = self.config.resman_options
step_options = resman_options.get(step['name'], [])

for config_key, config_val in step_options:
    flags += "-" if len(config_key) == 1 else "--"
    flags += f"{config_key}='{config_val}'"
```

the above also takes care of having values for options that have `=` in them

Reply:
I just realized that the above assumes that if `resman_options` is always assigned a value, which we can fix if we make the `PopperConfig` object always assign a dict, regardless of whether a config file is passed or not. In the latter case, the dict would be empty